### ISSUE TITLE
How to setup Kafka inside gradle-docker-compose-plugin?

### ISSUE THREAD

**Issue by iirekm** [avast/gradle-docker-compose-plugin#212]
```
Will use 192.168.16.1 (network test-local_default) as host of zookeeper
Will use 192.168.16.1 (network test-local_default) as host of kafka_1
Will use 192.168.16.1 (network test-local_default) as host of kafka_2
```

The plugin gives my IP address as container's address, which is wrong. The correct container address is the one under `Networks[0].IPAddress` returned by `docker inspect`, not the `Networks[0].Gateway`.

**Comment by augi** [comment]
Could you please describe in more details your use-case and why this behavior is wrong? Which platform do you use?

Btw. the output doesn't say that this is IP address of the container - it says that this is IP address that will be used to access the container. Then it also uses different ports.

According to our testing, usage of `Networks[0].Gateway` is the only reliable way how to make it working on all platforms, so Linux Containers, Linux Containers on Mac, Windows Containers, Linux Containers on Windows, ...

**Comment by iirekm** [comment]
Ubuntu bionic, a recent Docker CE package. I always used IP address for
discovering container ip in scripts and it worked fine. The address you
provide doesn't accept ports of the container.

My current workaround is exposing container ports to  docker host, but this
way parallel jenkins builds won't work 😔


śr., 27 lis 2019, 11:13 użytkownik Michal Augustýn <notifications@github.com>
napisał:

**Comment by augi** [comment]
The ports are mapped to another ports, the plugin reads them from `docker inspect`. The plugin offers these ports mapping in `dockerCompose.servicesInfos` property, and also allow to expose these ports as environment variables and Java System Properties.

**Comment by iirekm** [comment]
This port mapping needs at least this in .yml file:
```
 ports:
      - "0:19092"
```

It probably works for most things, but makes problems with Zookeeper and Kafka, where ports seem to be fixed. My current docker-compose.yml is:
```yaml
version: '2.4'

services:
  zookeeper:
    image: confluentinc/cp-zookeeper:5.3.1-1
    environment:
      ZOOKEEPER_CLIENT_PORT: 2181

  kafka_1:
    image: confluentinc/cp-kafka:5.3.1-1
    ports:
      - "19092:19092"
    depends_on:
      - zookeeper
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: PLAINTEXT://0.0.0.0:19092
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka_1:19092
      KAFKA_MESSAGE_MAX_BYTES: $KAFKA_REQUEST_SIZE
      KAFKA_REPLICA_FETCH_MAX_BYTES: $KAFKA_REQUEST_SIZE
      KAFKA_PRODUCER_MAX_REQUEST_SIZE: $KAFKA_REQUEST_SIZE
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 2
      KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 2

  kafka_2:
    image: confluentinc/cp-kafka:5.3.1-1
    ports:  
      - "19093:19093"
    depends_on:
      - zookeeper
    environment:
      KAFKA_BROKER_ID: 2
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: PLAINTEXT://0.0.0.0:19093
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka_2:19093
      KAFKA_MESSAGE_MAX_BYTES: $KAFKA_REQUEST_SIZE
      KAFKA_REPLICA_FETCH_MAX_BYTES: $KAFKA_REQUEST_SIZE
      KAFKA_PRODUCER_MAX_REQUEST_SIZE: $KAFKA_REQUEST_SIZE
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 2
      KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 2
```

Then `KAFKA_BOOTSTRAP_SERVERS=${KAFKA_1_HOST}:19092,${KAFKA_2_HOST}:19093` works.

This is ugly because requires hardcoding ports on host so parallel Jenkins jobs won't work. Whatever else I do (for example using `0:19092` mapping to get random port), I'm unable to connect with Kafka :-(

**Comment by augi** [comment]
Not sure what exactly you need but you should use `KAFKA_1_TCP_19092` instead of hardcoding `19092` (where `19092` is port inside the container - you don't have to assign an _external_ port).

So maybe your configuration should be something like `KAFKA_BOOTSTRAP_SERVERS=${KAFKA_1_HOST}:${KAFKA_1_TCP_19092},${KAFKA_2_HOST}:${KAFKA_2_TCP_19093}`?

**Comment by iirekm** [comment]
Tried - didn't help. I'm new to Kafka - there must be something wrong with
those fixed  KAFKA_LISTENERS: PLAINTEXT://0.0.0.0:19093
KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka_2:19093

śr., 27 lis 2019 o 15:12 Michal Augustýn <notifications@github.com>
napisał(a):

**Comment by augi** [comment]
This plugin just executes the `docker-compose up` command, there is no magic involved. So please make your Docker Compose environment working first, and involve this plugin later. I know that running Kafka in Docker Compose is very tricky.

I'm not able to help you with Kafka setup in Docker Compose, sorry :(
